spatialos-sdk/src/worker/schema.rs

```diff
@@ -41,3 +109,151 @@ pub trait SchemaPrimitiveField {
         result
     }
 }
+
+/// A struct that can be serialized into (and deserialized from) a [`SchemaObject`].
+///
+/// Schemalang [components] and [types], as well as the various associated types
+/// generated for components (i.e. updates and commands), are represented as a
+/// [`SchemaObject`] when serialized. This trait defines the conversion for such
+/// types, and provides a blanket implementation of the [`Field`] trait so that the
+/// types can be serialized as part of larger objects.
+///
+/// You should generally not have to implement this trait manually for any types.
+/// The implementation for any schema-defined types will be generated for you by the
+/// code generator provided with the SDK.
+///
+/// [components]: https://docs.improbable.io/reference/14.2/shared/schema/reference#components
+/// [types]: https://docs.improbable.io/reference/14.2/shared/schema/reference#types
+/// [`SchemaObject`]: struct.SchemaObject.html
+/// [`Field`]: trait.Field.html
+pub trait ObjectField: Sized + Clone {
+    /// Deserializes an instance of this type from a [`SchemaObject`].
+    ///
+    /// Note that deserialization can never fail as the underlying SpatialOS API ensures
+    /// that missing or invalid fields will always provide a valid default value during
+    /// deserialization. As such, if the data in `object` doesn't match what is expected
+    /// by the deserialization logic (e.g. due to a version mismatch between the current
+    /// schema version and the version `object` was generated from), the returned object
+    /// will have default values for any fields that are missing in `object` or have a
+    /// type differing from what is expected.
+    ///
+    /// [`SchemaObject`]: struct.SchemaObject.html
+    fn from_object(object: &SchemaObject) -> Self;
+
+    /// Serializes an instance of this type into a [`SchemaObject`].
+    ///
+    /// Note that serialization can never fail, as there is no way to form an invalid
+    /// instance of a schema object type.
+    ///
+    /// [`SchemaObject`]: struct.SchemaObject.html
+    fn into_object(&self, object: &mut SchemaObject);
+}
+
+impl<T> Field for T
+where
+    T: ObjectField,
+{
+    type RustType = Self;
+
+    fn get_or_default(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        Self::from_object(object.get_object(field))
+    }
+
+    fn index(object: &SchemaObject, field: FieldId, index: usize) -> Self::RustType {
+        Self::from_object(object.index_object(field, index))
+    }
+
+    fn count(object: &SchemaObject, field: FieldId) -> usize {
+        object.object_count(field)
+    }
+
+    fn add(object: &mut SchemaObject, field: FieldId, value: &Self::RustType) {
+        value.into_object(object.add_object(field));
+    }
+}
+
+/// The error type returned when a conversion to a schema enum type fails.
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct UnknownDiscriminantError;
```

Comment:
Should this have a field specifying what the schema value was when it failed? Could aid in debugging if this error ever arose

I.e. - 

```rust
pub struct UnknownDiscriminantError(u32);
```

Reply:
I considered that, but in any case where you're calling the `TryFrom`, you'll already have the invalid value; You're returning the unknown discriminant back to the code that gave it to you in the first place. That said, I'm not sure there's much of a downside to including the value in the error anyway, maybe just some missed opportunities for optimization since the error won't be a zero sized type anymore?

I'll add that in, since you're right about it possibly being useful in debugging (i.e. you can log the error object and it'll include the unknown discriminant).

Reply:
Okay, I did you one better and also added the type name to the error context, that way you can get actually useful error messages like:

```
Found unknown discriminant value 7 for enum `my_crate::schema::Foo` while deserializing schema data
```

Type names are `&'static str`, so it doesn't require any allocation or nothing :tada: